drivers/spi/spi_mcux_flexio.c

```diff
@@ -0,0 +1,481 @@
+		LOG_ERR("Half-duplex not supported");
+		return -ENOTSUP;
+	}
+
+	if (SPI_OP_MODE_GET(spi_cfg->operation) != SPI_OP_MODE_MASTER) {
+		LOG_ERR("Mode Slave not supported");
+		return -ENOTSUP;
+	}
+
+	FLEXIO_SPI_MasterGetDefaultConfig(&master_config);
+
+	word_size = SPI_WORD_SIZE_GET(spi_cfg->operation);
+	if ((word_size != 8) && (word_size != 16) && (word_size != 32)) {
+		LOG_ERR("Word size %d must be 8, 16 or 32", word_size);
+		return -EINVAL;
+	}
+	master_config.dataMode = word_size;
+
+	if (spi_cfg->operation & SPI_TRANSFER_LSB) {
+		if (word_size == 8) {
+			data->transfer_flags = kFLEXIO_SPI_8bitLsb;
+		} else if (word_size == 16) {
+			data->transfer_flags = kFLEXIO_SPI_16bitLsb;
+		} else {
+			data->transfer_flags = kFLEXIO_SPI_32bitLsb;
+		}
+	} else {
+		if (word_size == 8) {
+			data->transfer_flags = kFLEXIO_SPI_8bitMsb;
+		} else if (word_size == 16) {
+			data->transfer_flags = kFLEXIO_SPI_16bitMsb;
+		} else {
+			data->transfer_flags = kFLEXIO_SPI_32bitMsb;
+		}
+	}
+
+	if (flexio_api->get_rate(config->flexio_dev, &clock_freq)) {
+		return -EINVAL;
+	}
+
+#if (USE_FSL_HAL_FOR_INIT != 0)
+	master_config.dataMode = (word_size == 8)
+		? kFLEXIO_SPI_8BitMode
+		: kFLEXIO_SPI_16BitMode;
+
+	/* No CPOL for FlexIO_SPI */
+
+	master_config.phase =
+		(SPI_MODE_GET(spi_cfg->operation) & SPI_MODE_CPHA)
+		? kFLEXIO_SPI_ClockPhaseSecondEdge
+		: kFLEXIO_SPI_ClockPhaseFirstEdge;
+
+	master_config.baudRate_Bps = spi_cfg->frequency;
+
+	FLEXIO_SPI_MasterInit(config->flexio_spi, &master_config, clock_freq);
+#else
+	master_config.phase =
+		(SPI_MODE_GET(spi_cfg->operation) & SPI_MODE_CPHA)
+		? kFLEXIO_SPI_ClockPhaseSecondEdge
+		: kFLEXIO_SPI_ClockPhaseFirstEdge;
+
+	master_config.baudRate_Bps = spi_cfg->frequency;
+	spi_flexio_master_init(config->flexio_spi, &master_config,
+		(SPI_MODE_GET(spi_cfg->operation) & SPI_MODE_CPOL), clock_freq);
+#endif
+
+	FLEXIO_SPI_MasterTransferCreateHandle(config->flexio_spi, &data->handle,
+					spi_mcux_master_transfer_callback,
+					data);
+	/* No SetDummyData() for FlexIO_SPI */
+
+	data->ctx.config = spi_cfg;
+
+	return 0;
+}
+
+
+static int transceive(const struct device *dev,
+			const struct spi_config *spi_cfg,
+			const struct spi_buf_set *tx_bufs,
+			const struct spi_buf_set *rx_bufs,
+			bool asynchronous,
+			spi_callback_t cb,
+			void *userdata)
+{
+	const struct spi_mcux_flexio_config *config = dev->config;
+	const struct mcux_flexio_api *flexio_api = config->flexio_dev->api;
+	struct spi_mcux_flexio_data *data = dev->data;
+	int ret;
+
+	spi_context_lock(&data->ctx, asynchronous, cb, userdata, spi_cfg);
+
+	flexio_api->lock(config->flexio_dev);
+	ret = spi_mcux_flexio_configure(dev, spi_cfg);
+	flexio_api->unlock(config->flexio_dev);
+	if (ret) {
+		goto out;
+	}
+
+	spi_context_buffers_setup(&data->ctx, tx_bufs, rx_bufs, 1);
+
+	spi_context_cs_control(&data->ctx, true);
+
+	flexio_api->lock(config->flexio_dev);
+	flexio_api->interrupt_disable(config->flexio_dev);
```

Comment:
@danieldegrasse, since the `modify-read-write` operation on the `SHIFTSIEN` register occurs in the `FLEXIO_SPI_MasterTransferNonBlocking()` function, which is called from both the user thread and the interrupt handler. Also, the `FLEXIO_SPI_MasterTransferHandleIRQ()` function changes this register from the interrupt handler.